sig-robotics/examples/Definition_of_cloud_robot_CRD_and_Operator/internal/controller/utils.go

```diff
@@ -0,0 +1,144 @@
+package controller
+
+import (
+	"strconv"
+	"sync"
+	"time"
+
+	kerrors "errors"
+	glog "log"
+
+	"github.com/gorilla/websocket"
+	cloudrobotkubeedgev1beta1 "github.com/ospp2023/api/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+type IDPool struct {
+	mutex sync.Mutex
+	ids   []uint
+}
+
+var SharedIDPool *IDPool
+
+const (
+	RobotNodeTagName = "edge.cloudrobot.kubeedge.io/robot"
+	MaxRobotNum      = 100
+)
+
+func init() {
+	SharedIDPool = NewIDPool(MaxRobotNum)
+}
+
+func NewIDPool(maxID uint) *IDPool {
+	ids := make([]uint, 0, maxID)
+	for i := uint(1); i < maxID; i++ {
+		ids = append(ids, i)
+	}
+	return &IDPool{ids: ids}
+}
+
+func (p *IDPool) Allocate() (uint, error) {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	if len(p.ids) == 0 {
+		glog.Println("Failed to allocate id ")
+		return 0, kerrors.New("no available id to allocate")
+	}
+
+	id := p.ids[0]
+	p.ids = p.ids[1:]
+
+	return id, nil
+}
+
+func (p *IDPool) Release(id uint) {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	p.ids = append(p.ids, id)
```

Comment:
The `IDPool` struct and its methods are used to manage a pool of IDs for robots. However, the `Release` method does not check if the ID being released is already in the pool, which could lead to duplicate IDs. This could cause issues when allocating IDs. Consider adding a check in the `Release` method to prevent duplicates.
```diff
func (p *IDPool) Release(id uint) {
	p.mutex.Lock()
	defer p.mutex.Unlock()

+	if !ContainsElement(p.ids, id) {
+		p.ids = append(p.ids, id)
+	}
}
```